### PR TITLE
added summed dissimilarity sampler

### PR DESCRIPTION
**Pull request by musslick**
## Description

A simple (summed) dissimilarity sampler that samples N experiment conditions from a pool of X, which are most distant to data points in a reference pool X_base. The absolute distance for a given condition in X amounts to the sum of distances to each point in X_base.

### Type of change:
- New feature (non-breaking change which adds functionality)

### Features: 
- simple summed distance sampler
- unit test for 1-dimensional X
- unit test for N-dimensional X

### CODE REVIEW

**Review by younesStrittmatter — APPROVED** [review]
looks good to me

File: autora/experimentalist/sampler/dissimilarity.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+        "sokalsneath",
+        "yule",
+    ]
+
+    if metric not in valid_metrics:
+        raise ValueError(
+            f"Unsupported metric: '{metric}'\n" f"Only {valid_metrics} is supported."
+        )
+    else:
+        dist = DistanceMetric.get_metric(metric)
```
Comment (hollandjg) [inline autora/experimentalist/sampler/dissimilarity.py:79]: can we get away with just doing a one liner here:
```suggestion
    dist = DistanceMetric.get_metric(metric)
```
Will that fail sufficiently loudly for us? 

Combined with the type hint in the function signature and the MyPy checker, that will mean that most people won't use it wrong, but a sufficiently clever person will be able to implement their own metric, if they so choose. (They can add a new metric to scikit-learn and automatically be able to import it here.)

File: autora/experimentalist/sampler/dissimilarity.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+    # create a list to store the summed distances for each row in matrix1
+    summed_distances = []
+
+    # loop over each row in matrix1
+    for i, row in enumerate(X):
```
Comment (hollandjg) [inline autora/experimentalist/sampler/dissimilarity.py:85]: I don't see you using `i` – could you do `for row in X:`?

**Review by hollandjg — COMMENTED** [review]
This looks great! I've got a couple of suggestions for simplifications but that's all.

File: autora/experimentalist/sampler/dissimilarity.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+
+    Args:
+        X: pool of IV conditions to evaluate dissimilarity
+        X_ref: reference pool of IV conditions
+        n: number of samples to select
+        metric: dissimilarity measure. Options: 'euclidean', 'manhattan', 'chebyshev',
+            'minkowski', 'wminkowski', 'seuclidean', 'mahalanobis', 'haversine',
+            'hamming', 'canberra', 'braycurtis', 'matching', 'jaccard', 'dice',
+            'kulsinski', 'rogerstanimoto', 'russellrao', 'sokalmichener',
+            'sokalsneath', 'yule'. See `sklearn.metrics.DistanceMetric` for more details.
```
Comment (hollandjg) [inline autora/experimentalist/sampler/dissimilarity.py:23]: this could be simplified – see above.

Comment (musslick) [reply autora/experimentalist/sampler/dissimilarity.py:23]: That sounds great!

**Comment by younesStrittmatter** [comment]
while working on a sampler, that I derived from this,  I noticed something very minor: The variable is called X_ref while in the docstring it is called X_base

**Comment by musslick** [comment]
> 

Thanks for noticing! Should be fixed now...

**Review by hollandjg — APPROVED** [review]
This looks great! I've put together an additional PR #177 which extends our ability to link to external documentation, including scikit-learn, which builds on this.

**Comment by younesStrittmatter** [comment]
I have one more thing (sorry for the late message, I found this while playing around and it is really minor):
When you calculate the sums, you first put the single values into an array and then sum over the array (row_distances). It would be easier to just calculate the distance and add them to a sum (you looping anyway, so no need to first put it into an array and then sum over the array)

File: autora/experimentalist/sampler/dissimilarity.py [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+    dist = DistanceMetric.get_metric(metric)
+
+    # create a list to store the summed distances for each row in matrix1
+    summed_distances = []
+
+    # loop over each row in first matrix
+    for row in X:
+        # calculate the distances between the current row in matrix1 and all other rows in matrix2
+
+        row_distances = []
```
Comment (younesStrittmatter) [inline autora/experimentalist/sampler/dissimilarity.py:84]: instead:
summed_distance = 0

File: autora/experimentalist/sampler/dissimilarity.py [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+        for X_ref_row in X_ref:
+
+            distance = dist.pairwise([row, X_ref_row])[0, 1]
+            row_distances.append(distance)
```
Comment (younesStrittmatter) [inline autora/experimentalist/sampler/dissimilarity.py:88]: summed_distance += distance

File: autora/experimentalist/sampler/dissimilarity.py [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+
+        # calculate the summed distance for the current row
+        summed_distance = np.sum(row_distances)
```
Comment (younesStrittmatter) [inline autora/experimentalist/sampler/dissimilarity.py:91]: remove this line

**Review by younesStrittmatter — COMMENTED** [review]
this might be just a little more efficient

**Comment by musslick** [comment]
> this might be just a little more efficient

Great catch, I implemented that!
